lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -168,36 +182,158 @@ data BalanceInsufficientError = BalanceInsufficientError
 --
 performSelection
     :: forall m. (HasCallStack, MonadRandom m)
-    => SelectionCriteria
-    -> m (Either SelectionError SelectionResult)
-performSelection SelectionCriteria {outputsToCover, utxoAvailable}
+    => (TokenMap -> Coin)
+        -- ^ A function which computes the minimum required Ada coins for a
+        -- particular output.
+    -> (SelectionResult [AssetId] -> Coin)
+        -- ^ A function which computes the extra cost corresponding to a given
+        -- selection. This function must not depends on the value of each change
+        -- output (hence the parameterization).
+    -> SelectionCriteria
+        -- ^ The selection goal we're trying to satify.
+    -> m (Either SelectionError (UTxOIndex, SelectionResult TokenBundle))
+performSelection minCoinValueFor costFor criteria
     | not (balanceRequired `leq` balanceAvailable) =
         pure $ Left $ BalanceInsufficient $ BalanceInsufficientError
             {balanceAvailable, balanceRequired}
-    | otherwise =
-        Right . mkResult <$> runSelection utxoAvailable balanceRequired
+    | otherwise = do
+        SelectionState{selected, leftover} <-
+            runSelection extraCoinSource utxoAvailable balanceRequired
+
+        let preSelection :: SelectionResult [AssetId]
+            preSelection = mkPreSelection (mkInputsSelected selected)
+
+        makeChangeRepeatedly preSelection leftover <&> \case
+            Nothing -> -- Running out of ada-only UTxO to pay for fee / deposits
+                Left UnableToConstructChange
+            Just result ->
+                Right result
   where
+    SelectionCriteria{outputsToCover, utxoAvailable, extraCoinSource} = criteria
+
+    -- Outputs require a minimum ada value (to avoid filling the blockchain with
+    -- worthless UTxOs, any UTxO is required to hold some Ada which gives it an
+    -- implicit value).
+    --
+    -- However, users do not typically specify a minimum ada value themselves.
+    -- One would rather send '10 Apple' and not '10 Apple & 1.2 Ada'. Therefore,
+    -- unless a coin value is explicitely specified, we do assign a coin value
+    -- manually for each non-ada output. That value is the minimum value
+    -- possible to make a particular output valid.
+    outputsConsidered :: NonEmpty TxOut
+    outputsConsidered = flip fmap outputsToCover $ \out ->
+        out { Tx.tokens = augmentBundle (Tx.tokens out) }
+      where
+        augmentBundle bundle =
+            if TokenBundle.getCoin bundle == Coin 0 then
+                bundle { coin = minCoinValueFor (view #tokens bundle) }
+            else
+                bundle
```

Comment:
> Question: What if the user has specified a non-zero coin value that is lower than the result of minCoinValueFor?

We ought to catch that and tell them they did wrong. I haven't added that check yet but it is indeed the right place to do it. However, instead of taking the max, we should error. We shouldn't change outputs that are manually specified by users, but tell them instead about their mistakes.

Reply:
In that case, I'm not clear why the responsibility for checking minimum ada values should be in the `performSelection` function.

What if we were to instead:

1. Move the responsibility for assigning/checking the minimum ada values of outputs to the caller of the `performSelection` function.
2. Make it a pre-condition of `performSelection` that every output already satisfies `minCoinValueFor`.

Advantages:

1. Testing `performSelection` becomes more straightforward, as we don't have to worry about the augmenting-outputs-with-ada behaviour.
2. We wouldn't need the distinction between `outputsToCover` and `outputsCovered`, and this field could disappear from the `SelectionResult` record, making the documentation more straightforward.
3. The augmenting-outputs-with-ada functionality, if placed in its own function elsewhere, can be tested in isolation. (Right now, it's hard to test, as it's embedded within `performSelection`.)